Add benchmarking suite

ugognw opened the pull request:
## Summary

Major changes:

- feature 1: Addition of benchmarking suite (as per #57 ) - `pyEQL.benchmark`
- feature 2: added CRC dataset

## Todos

- [ ] validate data
- [ ] scrutinize dataset format
- [ ] finalize benchmarking statistics format
- [ ] remove redundant tests
- [ ] add more granular unit tests
- [ ] optimize number of test cases/time
- [ ] fix PhreeqcEOS benchmarking
- [ ] add docs under [Electrolyte Modeling Engines](https://pyeql.readthedocs.io/en/latest/engines.html)
- [ ] remove logic related to `_get_activity_coefficient_pitzer`?
- [ ] convert % concentration in datasets to mol/L in load_dataset
- [ ] Update changelog

- [X] Google format doc strings added.
- [X] Code linted with `ruff`. (For guidance in fixing rule violates, see [rule list](https://beta.ruff.rs/docs/rules/))
- [X] Type annotations included. Check with `mypy`.
- [X] Tests added for new features/fixes.
- [X] I have run the tests locally and they passed.

Happy to hear any feedback on the current direction!

ugognw committed:
Draft benchmarking module

Signed-off-by: Ugochukwu Nwosu <ugn@sfu.ca>

ugognw committed:
Major refactor

Replace ReferenceData with _BenchmarkEntry. This named tuple now separates solution
and solute properties which enables the separate retrieval of each type of property.
An analogous _get_solution_property method has been implemented.

_get_reference is renamed to _get_dataset as the data structure returned by the method
defines a general solution/solute property data set (reference data + Solution object
which can be used for benchmarking).

benchmark_engine functionality moved to report_results for simplicity. If more fine-tuned
control over data set partitioning or result reporting is desired, this can be implemented
quite readily by undoing this change.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Don't explicitly create soln; drafted CRC loader

Solution creation is now handed when reference files are read.

Benchmarking is parametrized by engine.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Draft parsers for benchmarking data sources

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Implement function to calculate mean activity

Although EOS subclasses define get_activity_coefficicient to return the
activity coefficients of individual solutes, one can only measure the
mean activity coefficient experimentally. As such, to benchmark against
experimental data, we must use the definition of the mean activity
coefficient in terms of the activity coefficients of the individual ions.
Alternatively, this function could be implemented as a property
(Solution.mean_activity) or a method (Solution.get_mean_activity).

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Clean up functions; add docstrings

Publicize functions/classes for processing input/output data models

Return values from main function (renamed to benchmark_engine for clarity) and pass
engine and source as arguments.

Store results instead of printing to console.

Parse Solution dictionary into Solution object when reading dataset from file.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix type hints

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix docstrings and errors

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Move reference data loading to own module

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Sketch loading functions

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Remove parsers

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Remove todos

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Optimized strategy-switching algorithm for sources

Passing a string corresponding to an internal source name retrieves the Path to
the data file (distributed with package). All files are loaded in the same way.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Parse dataset values into Quantity objects

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Draft benchmark tests

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Convert solute and solution data values

...to list[tuple[str, Quantity]]

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Rename property to match conventional name

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Expand benchmark unit tests

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix property retrieval, RMSE calc; rename vars

The new RMSE is more robust in the case that property values are Python
primitives.

Property getters are now called to retrieve properties. This fixes a bug
where the getter was returned instead of the value.

Variables are renamed (from property) to avoid shadowing a Python built-in.

Enumerating solution data (in get_dataset) yields the right number of items
to unpack.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix implementation of RMSE

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add molar conductivity reference data

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Rename function for clarity

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add utility to create BenchmarkEntry

Improve docstrings

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Remove extra Salt key; fix reduce signature

Salt takes no extra kwargs. reduce takes no kwargs.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Create source fixture

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add more properties

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add CRC DB files

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Rename function

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Update CRC data

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Remove empty file

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix type hints

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

rkingsbury commented:
Wow, thank you for the effort on this @ugognw! It looks well thought out. I'm traveling this week and part of next week but will dedicate some time to reviewing in detail as soon as I can.

ugognw committed:
Write new signature for benchmark_engine

Users should be able to select which properties as well as which solutions
(compositions, thermodynamic state) to benchmark against. This will also
enable speed-ups in testing.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add solns/props in load_dataset/BenchmarkEntry

Implementing BenchmarkEntry.solute_properties/solution_properties in this way
will enable speed-ups in testing when calculating error statistics between
calculated and reference datasets (list[BenchmarkEntry]) as data lookup from a
dictionary by key has O(1) time complexity (one will have the name of the property
when calculating error) while finding a specific item in a list has O(N) time
complexity.

Accordingly, load_dataset now expects solute/solution properties within data files
to be structured as dictionaries instead of prperty-value 2-tuples and accepts solutions,
solute_properties, and solution_properties as keyword-only arguments. That data for a
solution is included is determined by its components, molar concentration, temperature,
and pressure.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Defer decision on key structure

Encapsulating the logic for creating dictionary keys in a function and creating
a type alias for its type should minimize the amount of refactor required if the
format is changed and appease mypy. This also makes comparison of Solutions
for the purpose of benchmarking considerably more simple.

load_dataset now returns a dictionary with keys indicating the Solution for which
the data applies as generated by _create_solution_key.

Internal data sources are now opened more reliably using resources.files.

Generating the list of SolutionKeys corresponding to the `solutions` parameter in advance
simplifies the logic required to check for the Solution in `solutions`.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Use same data structures for engine/reference data

Using the same data structures for engine and reference data is desired because
it will separate the logic for retrieving properties and calculating stats and
simplify the retrieval of corresponding items in calculate_stats.

This implementation ensures that only one Solution object is created for each
Solution across all benchmark datasets and that each BenchmarkEntry object is
associated with a unique Solution. In this way, a single dataset can be generated
for the engine, which can then be benchmarked against each reference dataset.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Implement reference/calculated for calculate_stats

Accepting two identical data structures generalizes calculate_stats for
all combinations of sources of data (e.g., reference-reference, engine-engine,
engine-reference). An additional benefit is that now, from the signature, it
is more clear to users how to organize the two sets of data to be compared.
Previously, engine data was implicitly generated using the Solution objects in
each BenchmarkEntry. As mentioned previously, this now removes the logic for
retrieving solute and solution properties.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Reorder functions in order or use

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Update docstrings; remove unneeded func

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Remove database file

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Update database files

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Update tests; shrink solution properties tested

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Remove use of FormulaDict; fix bugs

FormulaDict was initially used in order to automatically standardize how solutes
appear in solution compositions and solute property lists. However, in addition
to standardizing ion formulas, FormulaDict also sorts its internal copy of the
dictionary by its values. This led to an error in the previous implementation
because the values are dictionaries, and dictionaries don't implement the required
methods for comparison with '<' and '>'.

Instead, we manually call standardize_formula in the public functions where
malformatted formulas are likely. This is when loading datasets and when
creating a BenchmarkEntry from a Solution.

Bugs: initialize property data as dictionaries in create_entry; fix indices and
data type in _create_engine_dataset; initialize data pair lists in calculate_stats.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Polish docstrings

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add units for concentration in SolutionKey

The concentration for Solution objects that are created in _create_engine_dataset
is determined from the SolutionKey. Without units, the amounts are not interpreted
as concentrations and the Solution does not have the correct composition.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add more unit tests

The three load_dataset unit tests are just outlines and are expected to fail.
The explicit test cases for test_should_benchmark_all_engines are intended to
eventually overlap with those in test_conductivity and eventually replace them.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Update module docstring

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Exclude the solvent from the solution key

The number of moles of solvent is dependent on the engine, but in the
capacity that SolutionKey will be used, two Solutions with the same concentrations
of their solutes with differing engines (and thus differing moles of solvent) should
be considered identical if they are at the same state (temperature and pressure).

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Outline additional unit tests

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix language in docstrings

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Revise docstrings

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Update docstrings; Fix metric args

Updated module docstring variable for consistency.

Require both _create_engine_dataset arguments.

For now, only require that the metric argument to calculate_stats can handle
Quantity arguments, so that we can ensure that units are handled correctly.
Note that the property values in BenchmarkEntry.solute_data and
BenchmarkEntry.solution_data are expected to be Quantity objects.

Also, the additional conditional statements in calculate_stats serve to handle
the edge cases of when the return values of _get_solute_property and
_get_solution_property are None and when there does not exist data in `calculated`
corresponding to `reference`. Because this may be of intereste to the user, the
event is logged at the INFO level.

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Fix typo

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw committed:
Add example for load_dataset

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

ugognw commented:
Thanks! I'd like to think so! Sounds good. I look forward to hearing your feedback!

ugognw committed:
Add benchmarking to docs

Signed-off-by: Ugochukwu Nwosu <ugognw@gmail.com>

rkingsbury commented:
> Wow, thank you for the effort on this @ugognw! It looks well thought out. I'm traveling this week and part of next week but will dedicate some time to reviewing in detail as soon as I can.

Update: I will do my best to look at this next week. I have some proposals due at the end of this week that will keep me busy until then. Thanks for your patience!

rkingsbury committed:
Merge branch 'main' into benchmarking-suite

rkingsbury committed:
Merge branch 'main' into benchmarking-suite

rkingsbury committed:
Merge branch 'main' into benchmarking-suite

rkingsbury commented:
Hi @ugognw , members of my lab group are going to continue work in this direction in #399 . If you're available and inclined to contributed, we welcome that, but no pressure.
